### PR TITLE
math: trig: Optimize with HiFi3 intrinsics the sin() and cos()

### PR DESCRIPTION
**Pull request by singalsu**
This patch optimizes for speed the sin() and cos() inline functions and the core cordic_sin_cos() function with Xtensa HiFI3 intrinsincs.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

### CODE REVIEW

**Comment by singalsu** [comment]
Please ignore the first commit, I need it to build cmocka. The beef is in the second commit.

**Review by lgirdwood — APPROVED** [review]
LGTM, @johnylin76 any comments ?

File: src/include/sof/math/trig.h [hunk 1]
```diff
@@ -12,6 +12,14 @@
 
 #include <stdint.h>
 
+#if defined __XCC__
+#include <xtensa/config/core-isa.h>
+#if XCHAL_HAVE_HIFI3 == 1
+#define CORDIC_HIFI3
```
Comment (cujomalainey) [inline src/include/sof/math/trig.h:18]: why define your own rather than use the provided define?

File: src/math/trig.c [hunk 2]
```diff
@@ -100,3 +166,5 @@ void cordic_sin_cos(int32_t th_rad_fxp, cordic_cfg type, int32_t *sign, int32_t
 }
 
 #endif
+
+#endif
```
Comment (cujomalainey) [inline src/math/trig.c:170]: add comments here so its clear what `if` this is closing please

File: test/cmocka/src/common_mocks.c [hunk 3]
```diff
@@ -97,6 +97,7 @@ int WEAK rstrlen(const char *s)
 void WEAK __panic(uint32_t p, char *filename, uint32_t linenum)
 {
 	fail_msg("panic: %s:%d (code 0x%x)\n", filename, linenum, p);
+	exit(1);
```
Comment (cujomalainey) [inline test/cmocka/src/common_mocks.c:100]: didnt this land in another PR? if yes, please rebase to remove from the diff

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

File: src/include/sof/math/trig.h (hunk 1 above)
Comment (singalsu) [reply src/include/sof/math/trig.h:18]: Umm, do yo mean use of XCHAL_HAVE_HIFI3 directly? I was thinking CORDIC_HIFI3 as generic HiFi3 compatible code enabler. While we don't have HiFi4 version done yet this could be set also for build in a HiFi4 platform due to backwards compatibility.

File: src/math/trig.c [hunk 4]
```diff
@@ -36,6 +36,72 @@ const int32_t cord_sincos_piovertwo_q29fl  = Q_CONVERT_FLOAT(_M_PI / 2, 29);
...
+		if ((th_rad_fxp + cord_sincos_piovertwo_q29fl) >= -cord_sincos_piovertwo_q28fl) {
+			th_rad_fxp += cord_sincos_piovertwo_q29fl;
+			*sign  = -1;
+		} else {
+			th_rad_fxp += cordic_sine_cos_piovertwo_q30fl;
+		}
+	}
+
+	th_rad_fxp = AE_SLAI32(th_rad_fxp, 2);
+	yn_xn = AE_SEL32_LL(0, cordic_sine_cos_lut_q29fl);
```
Comment (ShriramShastry) [inline src/math/trig.c:75]: Can you please confirm `Intrinsic AE_SEL32LL does not exist in the current configuration.` ?

Comment (ShriramShastry) [reply src/math/trig.c:75]: I check the implementation using hifi_bd5

File: src/math/trig.c [hunk 5]
```diff
@@ -36,6 +36,72 @@ const int32_t cord_sincos_piovertwo_q29fl  = Q_CONVERT_FLOAT(_M_PI / 2, 29);
...
+			th_rad_fxp += cordic_lookup[b_idx];
+			yn_xn = AE_SUBADD32(yn_xn, xy_tmp);
+		} else {
+			th_rad_fxp -= cordic_lookup[b_idx];
+			yn_xn = AE_ADDSUB32(yn_xn, xy_tmp);
+		}
+		xy_tmp = AE_SRAI32(AE_SRAA32(AE_SEL32_LH(yn_xn, yn_xn), b_idx), 1);
+	}
+
+	*b_yn = AE_SEL32_HH(yn_xn, yn_xn); /* TODO: 32bit store for high part? */
```
Comment (ShriramShastry) [inline src/math/trig.c:96]: I could not get `Intrinsic AE_SEL32HH does not exist in the current configuration.`

File: src/math/trig.c [hunk 6]
```diff
@@ -36,6 +36,72 @@ const int32_t cord_sincos_piovertwo_q29fl  = Q_CONVERT_FLOAT(_M_PI / 2, 29);
...
+	AE_S32_L_I(yn_xn, (ae_int32 *)xn, 0);
```
Comment (ShriramShastry) [inline src/math/trig.c:97]: I could not find `Intrinsic AE_S32LI does not exist in the current configuration`.

File: src/math/trig.c [hunk 7]
```diff
@@ -36,6 +36,72 @@ const int32_t cord_sincos_piovertwo_q29fl  = Q_CONVERT_FLOAT(_M_PI / 2, 29);
...
+	 */
+	for (b_idx = 0; b_idx < a_idx; b_idx++) {
+		if (th_rad_fxp < 0) {
+			th_rad_fxp += cordic_lookup[b_idx];
+			yn_xn = AE_SUBADD32(yn_xn, xy_tmp);
+		} else {
+			th_rad_fxp -= cordic_lookup[b_idx];
+			yn_xn = AE_ADDSUB32(yn_xn, xy_tmp);
+		}
+		xy_tmp = AE_SRAI32(AE_SRAA32(AE_SEL32_LH(yn_xn, yn_xn), b_idx), 1);
```
Comment (ShriramShastry) [inline src/math/trig.c:93]: I can't find `Intrinsic AE_SEL32LH does not exist in the current configuration`

File: src/math/trig.c (hunk 4 above)
Comment (singalsu) [reply src/math/trig.c:75]: You should use the core configurations for Intel APL, CNL, TGL, etc. platforms. In CI build of this PR passed for all platforms.

File: src/math/trig.c (hunk 2 above)
Comment (singalsu) [reply src/math/trig.c:170]: Yep, done!

File: test/cmocka/src/common_mocks.c (hunk 3 above)
Comment (singalsu) [reply test/cmocka/src/common_mocks.c]: Done.

**Comment by singalsu** [comment]
I'll keep this as draft until @ShriramShastry 's #4240 is merged. I will rebase this since the PR changes some functions those I edited here.

**Comment by singalsu** [comment]
Seems that codestyle / checkpatch check does not understand about xt-xcc AE integer types...

**Comment by singalsu** [comment]
@ShriramShastry I had to change cmpx_exp_xxb() functions as well since they use the same cordic_approx() as sine and cosine. There's no cmocka for it so the change is untested.  I'll keep PR as draft until we can test this.

File: src/math/trig.c [hunk 8]
```diff
@@ -270,6 +330,22 @@ int32_t is_scalar_cordic_asin(int32_t sinvalue, int16_t numiters)
  *		  struct  cordic_cmpx
  * Return Type	: none
  */
+#ifdef CORDIC_HIFI3
+void cmpx_cexp(ae_int32 sign, ae_int32 b_yn, ae_int32 xn, cordic_cfg type, struct cordic_cmpx *cexp)
+{
+	ae_int32 s;
+
+	cexp->re = sign * xn;
+	cexp->im = sign * b_yn;
```
Comment (singalsu) [inline src/math/trig.c:339]: This might be possible as dual multiply

Comment (singalsu) [reply src/math/trig.c:339]: Seems feasible only for loop since SIMD setup to ae_f32x2 and return to complex number placeholder struct takes more cycles.

File: src/math/trig.c [hunk 9]
```diff
@@ -100,6 +159,7 @@ void cordic_approx(int32_t th_rad_fxp, int32_t a_idx, int32_t *sign, int32_t *b_
 	/* Q2.30 format -sine, cosine*/
 	*th_cdc_fxp = th_rad_fxp;
 }
+#endif /* Generic C version */
```
Comment (cujomalainey) [inline src/math/trig.c:162]: The comment should match the if i.e. CORDIC_HIFI3 so its easy to match when viewing the brackets

File: src/math/trig.c (hunk 2 above)
Comment (cujomalainey) [reply src/math/trig.c:361]: is it pushed?

**Comment by cujomalainey** [comment]
ping on this review

File: src/math/trig.c (hunk 5 above)
Comment (singalsu) [reply src/math/trig.c]: You should use configuration for APL, CNL, TGL, etc.

File: src/math/trig.c (hunk 6 above)
Comment (singalsu) [reply src/math/trig.c]: ditto

File: src/math/trig.c (hunk 7 above)
Comment (singalsu) [reply src/math/trig.c:95]: ditto

File: src/include/sof/math/trig.h (hunk 1 above)
Comment (cujomalainey) [reply src/include/sof/math/trig.h:19]: Rather than going around and writting `|| XCHAL_HAVE_HIFI4` makes sens to me

File: src/include/sof/math/trig.h [hunk 10]
```diff
@@ -13,6 +13,14 @@
...
+#include <xtensa/tie/xt_hifi3.h>
+#endif
+#endif
```
Comment (cujomalainey) [inline src/include/sof/math/trig.h:22]: it would be good to have these `endifs` commented simply because they are nested

**Review by cujomalainey — APPROVED** [review]

File: src/include/sof/math/trig.h (hunk 10 above)
Comment (singalsu) [reply src/include/sof/math/trig.h]: Yep, I just updated these two #endif lines.

File: src/include/sof/math/trig.h [hunk 11]
```diff
@@ -89,18 +123,35 @@ static inline int32_t sin_fixed_32b(int32_t th_rad_fxp)
...
+	ae_int32 xn;
+	ae_int32 sign;
+
+	cordic_approx(th_rad_fxp, CORDIC_31B_TABLE_SIZE, &sign, &b_yn, &xn, &th_cdc_fxp);
+	th_cdc_fxp = sign * xn;
+
+	/*convert Q2.30 to Q1.31 format*/
+	return (ae_int32)AE_SLAI32S(th_cdc_fxp, 1);
+}
+#else
```
Comment (paulstelian97) [inline src/include/sof/math/trig.h:140]: In this specific situation, is the return statement the only real change? Feels like some unwarranted code duplication to me.

File: src/include/sof/math/trig.h [hunk 12]
```diff
@@ -118,6 +169,23 @@ static inline int32_t cos_fixed_32b(int32_t th_rad_fxp)
...
+	ae_int32 xn;
+	ae_int32 s;
+	ae_int32 sign;
+
+	cordic_approx(th_rad_fxp, CORDIC_15B_TABLE_SIZE, &sign, &b_yn, &xn, &th_cdc_fxp);
+	th_cdc_fxp = sign * b_yn;
+
+	/* convert Q2.30 to Q1.31 format, saturate to Q1.15*/
+	s = AE_SLAI32S(th_cdc_fxp, 1);
+	return AE_ROUND16X4F32SSYM(s, s);
```
Comment (paulstelian97) [inline src/include/sof/math/trig.h:186]: Ditto here.

**Comment by johnylin76** [comment]
Sorry for the late review. I have collected some statistics of speed along with the optimization of cordic functions.
The following MCPS data is measureeed on Eldrid (TGL) for the 3-band DRC of the internal speaker. I used tgl-013-drop-stable branch as the base with further adoptions of Malladi's and Seppo's commits on main branch.

MCPS       Scenario
127.476    without cordic functions (as is in tgl-012drop-stable)
165.000    with cordic sine 32-bit implementation
153.190    with cordic sine 16-bit implementation
143.980    with cordic sine and arc-sine 16-bit + Seppo's HiFi3 optimization
140.270    with cordic sine 16-bit + Seppo's HiFi3 optimization

Please also refer to [PR#4671](https://github.com/thesofproject/sof/pull/4671), thanks.

**Comment by singalsu** [comment]
> Sorry for the late review. I have collected some statistics of speed along with the optimization of cordic functions.
> The following MCPS data is measureeed on Eldrid (TGL) for the 3-band DRC of the internal speaker. I used tgl-013-drop-stable branch as the base with further adoptions of Malladi's and Seppo's commits on main branch.
> 
> MCPS Scenario
> 127.476 without cordic functions (as is in tgl-012drop-stable)
> 165.000 with cordic sine 32-bit implementation
> 153.190 with cordic sine 16-bit implementation
> 143.980 with cordic sine and arc-sine 16-bit + Seppo's HiFi3 optimization
> 140.270 with cordic sine 16-bit + Seppo's HiFi3 optimization
> 
> Please also refer to [PR#4671](https://github.com/thesofproject/sof/pull/4671), thanks.

Thanks for measurements @johny! I wonder if cordic 16 bit could be still optimized further, I'll take a look. There's the iteration loop and not much can be done to it unless there could be two parallel iterations (@ShriramShastry). I was hoping it to be closer to starting point MCPS. Now the benefit of cordic is smaller tables size but consume more cycles. Re-introduction of table lookup sine is also an option via kconfig choice.

File: src/include/sof/math/trig.h (hunk 11 above)
Comment (singalsu) [reply src/include/sof/math/trig.h:140]: Yep, the ae_int types are needed to avoid type casts.

File: src/math/trig.c [hunk 13]
```diff
@@ -43,6 +43,65 @@ const int32_t cord_arcsincos_q30fl  = Q_CONVERT_FLOAT(1.0000000000000000, 30);
...
+	/* Addition or subtraction by a multiple of pi/2 is done in the data type
+	 * of the input. When the fraction length is 29, then the quantization error
+	 * introduced by the addition or subtraction of pi/2 is done with 29 bits of
+	 * precision.Input range of cordicsin must be in the range [-2*pi, 2*pi),
+	 * a signed type with fractionLength = wordLength-4 will fit this range
+	 * without overflow.Increase of fractionLength makes the addition or
+	 * subtraction of a multiple of pi/2 more precise
+	 */
+	if (th_rad_fxp > cord_sincos_piovertwo_q28fl) {
+		if ((th_rad_fxp - cord_sincos_piovertwo_q29fl) <= cord_sincos_piovertwo_q28fl) {
```
Comment (singalsu) [inline src/math/trig.c:64]: This could be simplified to if (th_rad_fxp <= CONSTANT) where #define CONSTANT would be computed by cpp.

File: src/math/trig.c [hunk 14]
```diff
@@ -43,6 +43,65 @@ const int32_t cord_arcsincos_q30fl  = Q_CONVERT_FLOAT(1.0000000000000000, 30);
...
+			th_rad_fxp -= cord_sincos_piovertwo_q29fl;
+			*sign  = -1;
+		} else {
+			th_rad_fxp -= cordic_sine_cos_piovertwo_q30fl;
+		}
+	} else if (th_rad_fxp < -cord_sincos_piovertwo_q28fl) {
+		if ((th_rad_fxp + cord_sincos_piovertwo_q29fl) >= -cord_sincos_piovertwo_q28fl) {
```
Comment (singalsu) [inline src/math/trig.c:71]: same optimization if (th_rad_fxp >= CONSTANT2)

**Comment by lgirdwood** [comment]
> > Sorry for the late review. I have collected some statistics of speed along with the optimization of cordic functions.
> > The following MCPS data is measureeed on Eldrid (TGL) for the 3-band DRC of the internal speaker. I used tgl-013-drop-stable branch as the base with further adoptions of Malladi's and Seppo's commits on main branch.
> > MCPS Scenario
> > 127.476 without cordic functions (as is in tgl-012drop-stable)
> > 165.000 with cordic sine 32-bit implementation
> > 153.190 with cordic sine 16-bit implementation
> > 143.980 with cordic sine and arc-sine 16-bit + Seppo's HiFi3 optimization
> > 140.270 with cordic sine 16-bit + Seppo's HiFi3 optimization
> > Please also refer to [PR#4671](https://github.com/thesofproject/sof/pull/4671), thanks.
> 
> Thanks for measurements @johny! I wonder if cordic 16 bit could be still optimized further, I'll take a look. There's the iteration loop and not much can be done to it unless there could be two parallel iterations (@ShriramShastry). I was hoping it to be closer to starting point MCPS. Now the benefit of cordic is smaller tables size but consume more cycles. Re-introduction of table lookup sine is also an option via kconfig choice.

Yep, looks like a trade off between quality and performance (and should be select able in kconfig if the gap cant be closed).

**Comment by ShriramShastry** [comment]
> > > Sorry for the late review. I have collected some statistics of speed along with the optimization of cordic functions.
> > > The following MCPS data is measureeed on Eldrid (TGL) for the 3-band DRC of the internal speaker. I used tgl-013-drop-stable branch as the base with further adoptions of Malladi's and Seppo's commits on main branch.
> > > MCPS Scenario
> > > 127.476 without cordic functions (as is in tgl-012drop-stable)
> > > 165.000 with cordic sine 32-bit implementation
> > > 153.190 with cordic sine 16-bit implementation
> > > 143.980 with cordic sine and arc-sine 16-bit + Seppo's HiFi3 optimization
> > > 140.270 with cordic sine 16-bit + Seppo's HiFi3 optimization
> > > Please also refer to [PR#4671](https://github.com/thesofproject/sof/pull/4671), thanks.
> > 
> > 
> > Thanks for measurements @johny! I wonder if cordic 16 bit could be still optimized further, I'll take a look. There's the iteration loop and not much can be done to it unless there could be two parallel iterations (@ShriramShastry). I was hoping it to be closer to starting point MCPS. Now the benefit of cordic is smaller tables size but consume more cycles. Re-introduction of table lookup sine is also an option via kconfig choice.
> 
> Yep, looks like a trade off between quality and performance (and should be select able in kconfig if the gap cant be closed).
> > > Sorry for the late review. I have collected some statistics of speed along with the optimization of cordic functions.
> > > The following MCPS data is measureeed on Eldrid (TGL) for the 3-band DRC of the internal speaker. I used tgl-013-drop-stable branch as the base with further adoptions of Malladi's and Seppo's commits on main branch.
> > > MCPS Scenario
> > > 127.476 without cordic functions (as is in tgl-012drop-stable)
> > > 165.000 with cordic sine 32-bit implementation
> > > 153.190 with cordic sine 16-bit implementation
> > > 143.980 with cordic sine and arc-sine 16-bit + Seppo's HiFi3 optimization
> > > 140.270 with cordic sine 16-bit + Seppo's HiFi3 optimization
> > > Please also refer to [PR#4671](https://github.com/thesofproject/sof/pull/4671), thanks.
> > 
> > 
> > Thanks for measurements @johny! I wonder if cordic 16 bit could be still optimized further, I'll take a look. There's the iteration loop and not much can be done to it unless there could be two parallel iterations (@ShriramShastry). I was hoping it to be closer to starting point MCPS. Now the benefit of cordic is smaller tables size but consume more cycles. Re-introduction of table lookup sine is also an option via kconfig choice.
> 
> Yep, looks like a trade off between quality and performance (and should be select able in kconfig if the gap cant be closed).

There additional optimization which I think is missed in below section of code
```
	for (b_idx = 0; b_idx < a_idx; b_idx++) {
		if (th_rad_fxp < 0) {
			th_rad_fxp += cordic_lookup[b_idx];
			*xn += ytmp;
			*b_yn -= xtmp;
		} else {
			th_rad_fxp -= cordic_lookup[b_idx];
			*xn -= ytmp;
			*b_yn += xtmp;
		}
		xtmp = *xn >> (b_idx + 1);
		ytmp = *b_yn >> (b_idx + 1);
	}
```
I will include below section of code as a part of optimization ( branch predictor)

```
	for (b_idx = 0; b_idx < a_idx; b_idx++) {
		int32_t negate = (th_rad_fxp < 0) ? 1 : -1;
		th_rad_fxp += negate * cordic_lookup[b_idx];
		*xn += negate * ytmp;
		*b_yn -= negate * xtmp;
		xtmp = *xn >> (b_idx + 1);
		ytmp = *b_yn >> (b_idx + 1);
	}
```
I will include Seppo's suggestions and raise a new PR for improving the execution time.

**Comment by singalsu** [comment]
@ShriramShastry That should improve a bit. There's AE_LT32() operation that can be used to calculate negate.

**Comment by johnylin76** [comment]
To clarify that my MCPS data is measured upon the whole DRC (multiband_drc), which means sin() and asin() are just a part of the overall computation.

I have a thought to create a specified test firmware to iterate sin() or asin() several times running on TGL and measure the cpu ticks it spent. This measurement should be more valuable for optimization.

**Comment by singalsu** [comment]
I'll close this for now until #4694 is merged. The intrinsinc optimized version will be a bit different.

**Comment by johnylin76** [comment]
Hi Seppo and Malladi,
As I mentioned in the last comment, I have run some data for the cpu tick measurement of sin() implementations on TGL using perf_cnt tool. In brief I loop-run sin() 100 times with the pre-calculated input array uniformly distributed from pi/2 to -pi/2, then get the cpu tick from sof-logger and do the average.

CPU ticks per sin()   Implementation
69.92                       Legacy LUT
613.66                     Cordic 32bit generic
335.78                     Cordic 16bit generic
448.36                     Cordic 32bit HiFi3
217.96                     Cordic 16bit HiFi3

In fact from this perspective the legacy LUT implementation is still the fastest one. However we also could see that both 32b-to-16b and generic-to-HiFi3 do improve the cycle count a lot.

**Comment by ShriramShastry** [comment]
> Hi Seppo and Malladi,
> As I mentioned in the last comment, I have run some data for the cpu tick measurement of sin() implementations on TGL using perf_cnt tool. In brief I loop-run sin() 100 times with the pre-calculated input array uniformly distributed from pi/2 to -pi/2, then get the cpu tick from sof-logger and do the average.
> 
> CPU ticks per sin() Implementation
> 69.92 Legacy LUT
> 613.66 Cordic 32bit generic
> 335.78 Cordic 16bit generic
> 448.36 Cordic 32bit HiFi3
> 217.96 Cordic 16bit HiFi3
> 
> In fact from this perspective the legacy LUT implementation is still the fastest one. However we also could see that both 32b-to-16b and generic-to-HiFi3 do improve the cycle count a lot.

Thanks !! Does It means cordic sin 32- bit  to 16-bit implementation we have ~50% CPU ticks reduction ?

**Comment by singalsu** [comment]
> In fact from this perspective the legacy LUT implementation is still the fastest one.

@lgirdwood I think we should re-introduce it as opt-out in Kconfig. But I wonder if it should be called sin_lut() something. It has accuracy between 32 and 16 bit. I recall it's about 20 bits precision.
